### ISSUE TITLE
package provider action `:upgrade` should work with Ubuntu

### ISSUE THREAD

**Issue by haidangwa** [chef/chef#11103]
### Describe the Enhancement:

When I declare a `package` resource in my recipe, it is smart enough to determine that my chef node is an Ubuntu server. It then proceeds to use the `apt_package` resource to do an upgrade; however, the `upgrade` action is not supported by `apt_package`. Instead of intelligently changing the provider to `dpkg_package`, chef client fails and throws an error. Ironically, the error informs me that I should use the `dpkg_package` resource to perform an `upgrade` action.

The installer is downloaded via `remote_file` before I get to the package resource, declared below. My cookbook is used on ubuntu, rhel, oel, and amazon linux chef clients. So, I don't want to specify different `*_package` providers, like `yum_package` on rhel and `dpkg_package` on ubuntu with some guards. Seems rather clunky code. Instead, the package resource should be able to meet my needs, except it fails on ubuntu.

```
package 'falcon-sensor' do
  action :upgrade
  source falcon_sensor_pkg['local_file']
end
```

### Describe the Need:

It seems odd to me that the chef client doesn't just do this automatically for me -- use `apt_package` for most cases, but if `action :upgrade` is provided to the `package` resource, then it should automatically use the `dpkg_package` resource under the covers.

The chef run fails with this error:
```
           * apt package provider cannot handle source property. Use dpkg provider instead
           ================================================================================
           Error executing action `upgrade` on resource 'apt_package[falcon-sensor]'
           ================================================================================
           
           Chef::Exceptions::Package
           -------------------------
           apt package provider cannot handle source property. Use dpkg provider instead
           
           Resource Declaration:
           ---------------------
           # In /tmp/kitchen/cache/cookbooks/linux_base/recipes/falcon_sensor.rb
           
            44: package 'falcon-sensor' do
            45:   action :upgrade
            46:   source falcon_sensor_pkg['local_file']
            47: end
            48: 

```
The first line indicates to me that chef thinks I should use the dpkg provider instead to perform an upgrade.

### Current Alternative

I have to add a workaround to my package resource to explicitly give it the proper provider to perform an upgrade.

```
package 'falcon-sensor' do
  action :upgrade
  source falcon_sensor_pkg['local_file']
  provider Chef::Provider::Package::Dpkg if platform?('ubuntu')
end
```

**Comment by lamont-granquist** [comment]
So this should not be implemented as the apt_package resource calling dpkg or dpkg_package to do the install (we would need code in the apt package provider itself similar to yum's `resolve_source_to_version_obj` that calls dpkg to get the package version number).

This should be done through using the `use_package_name_for_source` subclass directive, and implemented using `apt -sourcepath` so that dependency resolution happens for the deb.  This is similar to how yum+rpm package providers work.

TL;DR:  make it work like yum_package does.
